File: THPTTHD/app/src/main/java/com/example/thpt_thd/MainActivity.java
package com.example.thpt_thd;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.NotificationCompat;


import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Intent;
import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.util.Log;
import android.util.Pair;
import android.view.View;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.TextView;
import android.widget.Toast;

import org.eclipse.paho.android.service.MqttAndroidClient;
import org.eclipse.paho.client.mqttv3.IMqttActionListener;
import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
import org.eclipse.paho.client.mqttv3.IMqttToken;
import org.eclipse.paho.client.mqttv3.MqttCallback;
import org.eclipse.paho.client.mqttv3.MqttClient;
import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
import org.eclipse.paho.client.mqttv3.MqttException;
import org.eclipse.paho.client.mqttv3.MqttMessage;

import java.nio.charset.StandardCharsets;
import java.util.Random;

public class MainActivity extends AppCompatActivity {
    private String serverURL = "tcp://test.mosquitto.org:1883";
    MqttAndroidClient client;
    RequestQueue requestQueue;
    private Button resolveButton;
    private TextView text;
    private CheckBox g1, g2, g3, g4;

    private WebView webView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        this.webView = findViewById(R.id.webView);

        this.requestQueue = new RequestQueue();
        setTitle(getString(R.string.schoolName));
        text = (TextView) this.findViewById(R.id.text);
        resolveButton = (Button) this.findViewById(R.id.button_accept);

        g1 = (CheckBox) findViewById(R.id.checkBox1);
        g2 = (CheckBox) findViewById(R.id.checkBox2);
        g3 = (CheckBox) findViewById(R.id.checkBox3);

        initMqtt(serverURL);

        resolveButton.setOnClickListener(v -> handleBtnResolveClick());

        setupOnClickListenerForCheckboxs();
    }

    protected void registerCameraViewer(String webURL) {
        WebSettings webSettings = webView.getSettings();
        webSettings.setJavaScriptEnabled(true);
        webSettings.setDomStorageEnabled(true);
        webView.setWebChromeClient(new WebChromeClient());
        webView.loadUrl(webURL);
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        setIntent(intent); // 🔹 Update the intent for this activity

        // 🔹 Handle the updated "room" value
        String room = intent.getStringExtra("room");
        if (room != null) {
            Log.d("MainActivity", "Relaunched with room: " + room);
            registerCameraViewer("https://rtsp.gasbinhminh.vn");
            Toast.makeText(this, "camera " + room, Toast.LENGTH_SHORT).show();
        }
    }

    private void setupOnClickListenerForCheckboxs() {
        g1.setOnClickListener(v -> {
            boolean checked = ((CheckBox) v).isChecked();
            // Check which checkbox was clicked
            if (checked) {
                sub("401");
            } else {
                unsub("401");
            }
        });

        g2.setOnClickListener(v -> {
            boolean checked = ((CheckBox) v).isChecked();
            // Check which checkbox was clicked
            if (checked) {
                sub("402");
            } else {
                unsub("402");
            }
        });

        g3.setOnClickListener(v -> {
            boolean checked = ((CheckBox) v).isChecked();
            // Check which checkbox was clicked
            if (checked) {
                sub("403");
            } else {
                unsub("403");
            }
        });
    }

    private void initMqtt(String serverURL) {
        String clientId = MqttClient.generateClientId();
        client = new MqttAndroidClient(this.getApplicationContext(), serverURL, clientId);
        client.setCallback(new MqttCallback() {
            @Override
            public void connectionLost(Throwable cause) {
                Log.d("mqtt", "connectionLost");
                Toast.makeText(getApplication(), "mqtt disconnected", Toast.LENGTH_SHORT).show();
            }

            @Override
            public void messageArrived(String topic, MqttMessage mqttMessage) {
                Log.d("mqtt", "messageArrived:" + mqttMessage.toString());
                handleMessageArrived(topic, mqttMessage.toString());
            }

            @Override
            public void deliveryComplete(IMqttDeliveryToken token) {
                Log.d("mqtt", "deliveryComplete");
            }
        });

        try {
            MqttConnectOptions options = new MqttConnectOptions();
            options.setAutomaticReconnect(true);
            IMqttToken token = client.connect(options);
            token.setActionCallback(new IMqttActionListener() {
                @Override
                public void onSuccess(IMqttToken asyncActionToken) {
                    Log.d("mqtt", "client connected");
                    Toast.makeText(getApplication(), "mqtt server connected: " + serverURL, Toast.LENGTH_LONG).show();
                }
                @Override
                public void onFailure(IMqttToken asyncActionToken, Throwable exception) {
                    Log.d("mqtt", "client failed to connect");
                    Toast.makeText(getApplication(), "mqtt client failed to connect", Toast.LENGTH_SHORT).show();
                }
            });
        } catch (MqttException e) {
            e.printStackTrace();
        }
    }

    private void handleMessageArrived(String room, String message) {
        // TODO: handle incoming messages
        switch (message) {
            case "Cap cuu": {
                requestQueue.add(room, true);
                sendNoti(room, message);
                break;
            }
            case "Ho tro": {
                requestQueue.add(room, false);
                sendNoti(room, message);
                break;
            }
            case "OK": {
                requestQueue.remove(room);
                // clearNoti(room);
                Toast.makeText(this, room + " đã được tiếp nhận", Toast.LENGTH_LONG).show();
                break;
            }
            default: {
                Log.d("mqtt", "warning received invalid message: " + message);
            }
        }
        updateAlarm();
    }

    private void handleBtnResolveClick() {
        if (requestQueue.isEmpty()) return;

        Pair<String, Boolean> victim = requestQueue.remove();
        String room = victim.first;
//        clearNoti(room);
        pub(room + "_re", "accept");
        pub(room, "OK");
        updateAlarm();
        Toast.makeText(this, "Đã tiếp nhận", Toast.LENGTH_SHORT).show();
    }

    /***
     * Manage displayed text and alarm sound
     */
    private void updateAlarm() {
        if (requestQueue.isEmpty()) {
            text.setText(R.string.textHolder);
            stopSound();
        } else {
            Pair<String, Boolean> req = requestQueue.peek();
            String room = req.first;
            Boolean isUrgent = req.second;
            if (isUrgent) {
                text.setText(room + ": Cấp cứu");
            } else {
                text.setText(room + ": Hỗ trợ");
            }
            playSound(isUrgent);
        }
    }

    void pub(String topic, String content) {
        byte[] encodedPayload;
        try {
            encodedPayload = content.getBytes(StandardCharsets.UTF_8);
            MqttMessage message = new MqttMessage(encodedPayload);
            client.publish(topic, message);
        } catch (MqttException e) {
            e.printStackTrace();
        }
        Log.d("mqtt", "pub");
        Toast.makeText(this, "message sent", Toast.LENGTH_SHORT).show();
    }

    void sub(String topic) {
        int qos = 0;
        try {
            IMqttToken subToken = client.subscribe(topic, qos);
            subToken.setActionCallback(new IMqttActionListener() {
                @Override
                public void onSuccess(IMqttToken asyncActionToken) {
                    // The message was published
                    Log.d("mqtt", "sub success");
                    Toast.makeText(getApplication(), "subscribed. topic " + topic, Toast.LENGTH_SHORT).show();
                }

                @Override
                public void onFailure(IMqttToken asyncActionToken,
                                      Throwable exception) {
                    Log.d("mqtt", "sub fail");
                    Toast.makeText(getApplication(), "failed to subscribe topic", Toast.LENGTH_SHORT).show();
                    // The subscription could not be performed, maybe the user was not
                    // authorized to subscribe on the specified topic e.g. using wildcards

                }
            });
        } catch (MqttException e) {
            e.printStackTrace();
        }
    }

    void unsub(String topic) {
        try {
            IMqttToken unsubToken = client.unsubscribe(topic);
            unsubToken.setActionCallback(new IMqttActionListener() {
                @Override
                public void onSuccess(IMqttToken asyncActionToken) {
                    // The subscription could successfully be removed from the client
                    Toast.makeText(getApplication(), "unsubscribed. topic " + topic, Toast.LENGTH_SHORT).show();
                }

                @Override
                public void onFailure(IMqttToken asyncActionToken,
                                      Throwable exception) {
                    Toast.makeText(getApplication(), "failed to unsubscribe topic", Toast.LENGTH_SHORT).show();
                    // some error occurred, this is very unlikely as even if the client
                    // did not had a subscription to the topic the unsubscribe action
                    // will be successfully
                }
            });
        } catch (MqttException e) {
            e.printStackTrace();
        }
    }

    void sendNoti(@NonNull String topic, @NonNull String msg) {
        // 🔹 Intent to bring MainActivity to the foreground
        Intent intent = new Intent(this, MainActivity.class);
        intent.putExtra("room", topic); // Pass the room as an extra
        intent.putExtra("isUrgent", true); // Pass the room as an extra
        intent.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP | Intent.FLAG_ACTIVITY_CLEAR_TOP);

        // 🔹 Wrap the intent in a PendingIntent
        PendingIntent pendingIntent = PendingIntent.getActivity(
            this,
            0,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE // Ensure it updates with new extras
        );

        Bitmap bitmap = BitmapFactory.decodeResource(getResources(), R.drawable.ic_launcher_foreground);
        @SuppressLint("WrongConstant") Notification noti = new NotificationCompat.Builder(this, MyApplication.CHANNEL_ID)
            .setContentTitle(topic)
            .setContentText(msg)
            .setSmallIcon(R.drawable.small)
            .setLargeIcon(bitmap)
            .setPriority(NotificationCompat.PRIORITY_DEFAULT)
            .setAutoCancel(true)
            .setContentIntent(pendingIntent)
            .setSilent(true)
            .build();

        int notificationId;
        try {
            notificationId = Integer.parseInt(topic);
        } catch (NumberFormatException e) {
            notificationId = 1;
        }

        NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
        if (notificationManager != null) {
            notificationManager.notify(notificationId, noti);
        }
    }

    void clearNoti(@NonNull String topic) {
        int notificationId;
        try {
            notificationId = Integer.parseInt(topic);
        } catch (NumberFormatException e) {
            notificationId = 1;
        }
        NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
        if (notificationManager != null) {
            notificationManager.cancel(notificationId);
        }
    }

    void playSound(boolean isUrgent) {
        stopSound();
        Context context = getApplicationContext();
        Intent startIntent = new Intent(context, RingtoneService.class);
        startIntent.putExtra("isUrgent", isUrgent);
        context.startService(startIntent);
    }

    void stopSound() {
        Context context = getApplicationContext();
        Intent stopIntent = new Intent(context, RingtoneService.class);
        context.stopService(stopIntent);
    }
}

File: THPTTHD/app/src/main/java/com/example/thpt_thd/MainActivity.kt
package com.example.thpt_thd

import android.annotation.SuppressLint
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Intent
import android.graphics.BitmapFactory
import android.os.Bundle
import android.util.Log
import android.view.View
import android.webkit.WebSettings
import android.widget.CheckBox
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.NotificationCompat
import com.example.thpt_thd.databinding.ActivityMainBinding
import org.eclipse.paho.android.service.MqttAndroidClient
import org.eclipse.paho.client.mqttv3.*
import java.nio.charset.StandardCharsets


class MainActivity : AppCompatActivity() {
    private var shouldAlarmShown = false
    private var client: MqttAndroidClient? = null
    private var requestQueue: RequestQueue = RequestQueue()

    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        title = getString(R.string.schoolName)

        binding.cardView1.roomText.text = "Phòng học 4A1"
        binding.cardView2.roomText.text = "Phòng học 4A2"
        binding.cardView3.roomText.text = "Phòng học 4A3"
        binding.cardView1.roomImage.setOnClickListener { showCameraViewer() }
        binding.cardView2.roomImage.setOnClickListener { showCameraViewer() }
        binding.cardView3.roomImage.setOnClickListener { showCameraViewer() }

        binding.alarmAcceptButton.setOnClickListener { handleBtnResolveClick() }
        binding.alarmWebView.visibility = View.INVISIBLE
        shouldAlarmShown = false
        refreshView(this.shouldAlarmShown)
        initMqtt("tcp://test.mosquitto.org:1883")
        setupOnClickListenerForCheckboxs()
    }

    private fun refreshView(shouldAlarmShown: Boolean) {
        binding.dashboard.visibility = if (!shouldAlarmShown) View.VISIBLE else View.INVISIBLE
        binding.alarmAcceptButton.visibility = if (shouldAlarmShown) View.VISIBLE else View.INVISIBLE
        binding.alarmText.visibility = if (shouldAlarmShown) View.VISIBLE else View.INVISIBLE
    }

    private fun showCameraViewer(webURL: String) {
        binding.alarmWebView.visibility = View.VISIBLE
        Toast.makeText(applicationContext, "loading..", Toast.LENGTH_SHORT).show()
        binding.alarmWebView.settings.javaScriptEnabled = true
        binding.alarmWebView.loadUrl(webURL)
    }

    private fun showCameraViewer() {
        val webURL = "https://rtsp.gasbinhminh.vn/stream"
        showCameraViewer(webURL)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent) // 🔹 Update the intent for this activity

        // 🔹 Handle the updated "room" value
        val room = intent.getStringExtra("room")
        if (room != null) {
            showCameraViewer()
            shouldAlarmShown = true
            refreshView(this.shouldAlarmShown)
        }
    }

    private fun setupOnClickListenerForCheckboxs() {
        binding.cardView1.subscribeCheckBox.setOnClickListener { v: View ->
            val checked = (v as CheckBox).isChecked
            // Check which checkbox was clicked
            if (checked) {
                sub("401")
            } else {
                unsub("401")
            }
        }
        binding.cardView2.subscribeCheckBox.setOnClickListener { v: View ->
            val checked = (v as CheckBox).isChecked
            // Check which checkbox was clicked
            if (checked) {
                sub("402")
            } else {
                unsub("402")
            }
        }
        binding.cardView3.subscribeCheckBox.setOnClickListener { v: View ->
            val checked = (v as CheckBox).isChecked
            // Check which checkbox was clicked
            if (checked) {
                sub("403")
            } else {
                unsub("403")
            }
        }
    }

    private fun initMqtt(serverURL: String) {
        val clientId = MqttClient.generateClientId()
        client = MqttAndroidClient(applicationContext, serverURL, clientId)
        client!!.setCallback(object : MqttCallback {
            override fun connectionLost(cause: Throwable) {
                Log.d("mqtt", "connectionLost")
                Toast.makeText(application, "mqtt disconnected", Toast.LENGTH_SHORT).show()
            }

            override fun messageArrived(topic: String, mqttMessage: MqttMessage) {
                Log.d("mqtt", "messageArrived:$mqttMessage")
                handleMessageArrived(topic, mqttMessage.toString())
            }

            override fun deliveryComplete(token: IMqttDeliveryToken) {
                Log.d("mqtt", "deliveryComplete")
            }
        })
        try {
            val options = MqttConnectOptions()
            options.isAutomaticReconnect = true
            val token = client!!.connect(options)
            token.actionCallback = object : IMqttActionListener {
                override fun onSuccess(asyncActionToken: IMqttToken) {
                    Log.d("mqtt", "client connected")
                    Toast.makeText(
                        application,
                        "mqtt server connected: $serverURL",
                        Toast.LENGTH_LONG
                    ).show()
                }

                override fun onFailure(asyncActionToken: IMqttToken, exception: Throwable) {
                    Log.d("mqtt", "client failed to connect")
                    Toast.makeText(application, "mqtt client failed to connect", Toast.LENGTH_SHORT)
                        .show()
                }
            }
        } catch (e: MqttException) {
            e.printStackTrace()
        }
    }

    private fun handleMessageArrived(room: String, message: String) {
        // TODO: handle incoming messages
        when (message) {
            "Cap cuu" -> {
                requestQueue.add(room, true)
                sendNoti(room, message)
            }
            "Ho tro" -> {
                requestQueue.add(room, false)
                sendNoti(room, message)
            }
            "OK" -> {
                requestQueue.remove(room)
                // clearNoti(room);
                Toast.makeText(this, "$room đã được tiếp nhận", Toast.LENGTH_LONG).show()
            }
            else -> {
                Log.d("mqtt", "warning received invalid message: $message")
            }
        }
        updateAlarm()
    }

    private fun handleBtnResolveClick() {
        if (!requestQueue.isEmpty) {
            val latestAlarm = requestQueue.remove()
            val room = latestAlarm.first
            pub(room, "OK")
        }
        shouldAlarmShown = false
        refreshView(this.shouldAlarmShown)
        updateAlarm()
    }

    /***
     * Manage displayed text and alarm sound
     */
    private fun updateAlarm() {
        if (requestQueue.isEmpty) {
            binding.alarmText.text = ""
            stopSound()
        } else {
            val req = requestQueue.peek()
            val room = req.first
            val isUrgent = req.second
            val msg = if (isUrgent) "$room: Tín hiệu khẩn cấp 🆘" else "$room: Cần hỗ trợ ✋"
            binding.alarmText.text = msg
            playSound(isUrgent)
        }
    }

    fun pub(topic: String?, content: String) {
        val encodedPayload: ByteArray
        try {
            encodedPayload = content.toByteArray(StandardCharsets.UTF_8)
            val message = MqttMessage(encodedPayload)
            client!!.publish(topic, message)
        } catch (e: MqttException) {
            e.printStackTrace()
        }
        Log.d("mqtt", "pub")
        Toast.makeText(this, "message sent", Toast.LENGTH_SHORT).show()
    }

    private fun sub(topic: String) {
        val qos = 0
        try {
            val subToken = client!!.subscribe(topic, qos)
            subToken.actionCallback = object : IMqttActionListener {
                override fun onSuccess(asyncActionToken: IMqttToken) {
                    // The message was published
                    Log.d("mqtt", "sub success")
                    Toast.makeText(application, "subscribed. topic $topic", Toast.LENGTH_SHORT)
                        .show()
                }

                override fun onFailure(
                    asyncActionToken: IMqttToken,
                    exception: Throwable
                ) {
                    Log.d("mqtt", "sub fail")
                    Toast.makeText(application, "failed to subscribe topic", Toast.LENGTH_SHORT)
                        .show()
                    // The subscription could not be performed, maybe the user was not
                    // authorized to subscribe on the specified topic e.g. using wildcards
                }
            }
        } catch (e: MqttException) {
            e.printStackTrace()
        }
    }

    private fun unsub(topic: String) {
        try {
            val unsubToken = client!!.unsubscribe(topic)
            unsubToken.actionCallback = object : IMqttActionListener {
                override fun onSuccess(asyncActionToken: IMqttToken) {
                    // The subscription could successfully be removed from the client
                    Toast.makeText(application, "unsubscribed. topic $topic", Toast.LENGTH_SHORT)
                        .show()
                }

                override fun onFailure(
                    asyncActionToken: IMqttToken,
                    exception: Throwable
                ) {
                    Toast.makeText(application, "failed to unsubscribe topic", Toast.LENGTH_SHORT)
                        .show()
                    // some error occurred, this is very unlikely as even if the client
                    // did not had a subscription to the topic the unsubscribe action
                    // will be successfully
                }
            }
        } catch (e: MqttException) {
            e.printStackTrace()
        }
    }

    fun sendNoti(topic: String, msg: String) {
        // 🔹 Intent to bring MainActivity to the foreground
        val intent = Intent(this, MainActivity::class.java)
        intent.putExtra("room", topic) // Pass the room as an extra
        intent.putExtra("isUrgent", true) // Pass the room as an extra
        intent.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP or Intent.FLAG_ACTIVITY_CLEAR_TOP

        // 🔹 Wrap the intent in a PendingIntent
        val pendingIntent = PendingIntent.getActivity(
            this,
            0,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE // Ensure it updates with new extras
        )
        val bitmap = BitmapFactory.decodeResource(resources, R.drawable.ic_launcher_foreground)
        @SuppressLint("WrongConstant") val noti =
            NotificationCompat.Builder(this, MyApplication.CHANNEL_ID)
                .setContentTitle(topic)
                .setContentText(msg)
                .setSmallIcon(R.drawable.small)
                .setLargeIcon(bitmap)
                .setPriority(NotificationCompat.PRIORITY_DEFAULT)
                .setAutoCancel(true)
                .setContentIntent(pendingIntent)
                .setSilent(true)
                .build()
        val notificationId: Int = try {
            topic.toInt()
        } catch (e: NumberFormatException) {
            1
        }
        val notificationManager = getSystemService(NOTIFICATION_SERVICE) as NotificationManager
        notificationManager.notify(notificationId, noti)
    }

    fun playSound(isUrgent: Boolean) {
        stopSound()
        val context = applicationContext
        val startIntent = Intent(context, RingtoneService::class.java)
        startIntent.putExtra("isUrgent", isUrgent)
        context.startService(startIntent)
    }

    fun stopSound() {
        val context = applicationContext
        val stopIntent = Intent(context, RingtoneService::class.java)
        context.stopService(stopIntent)
    }
}